File: src/data/mcp.ts
import { execSync, spawnSync } from 'child_process';
import { existsSync, readFileSync } from 'fs';
import { join } from 'path';
import { homedir } from 'os';
import type { OcMcpServer, McpDependency } from '../types.js';

const OPENCLAW_JSON = join(homedir(), '.openclaw', 'openclaw.json');

interface McpServerEntry {
  url?: string;
  command?: string;
  args?: string[];
  transport?: string;
  headers?: Record<string, string>;
  enabled?: boolean;
}

interface OpenclawJson {
  mcp?: {
    servers?: Record<string, McpServerEntry>;
  };
}

function checkCommand(cmd: string): boolean {
  try {
    const r = spawnSync('which', [cmd], { encoding: 'utf-8', timeout: 2000 });
    return r.status === 0;
  } catch {
    return false;
  }
}

function checkHttpUrl(url: string): boolean {
  try {
    const out = execSync(
      `curl --head --max-time 3 -s -o /dev/null -w "%{http_code}" ${JSON.stringify(url)}`,
      { encoding: 'utf-8', timeout: 5000 }
    ).trim();
    const code = parseInt(out, 10);
    // Any real HTTP response (even 401/403/404) means the server is reachable
    return code > 0;
  } catch {
    return false;
  }
}

export function loadMcp(): OcMcpServer[] {
  if (!existsSync(OPENCLAW_JSON)) return [];

  try {
    const config: OpenclawJson = JSON.parse(readFileSync(OPENCLAW_JSON, 'utf-8'));
    const servers = config.mcp?.servers ?? {};

    return Object.entries(servers).map(([name, entry]) => {
      const transport = entry.transport ?? (entry.command ? 'stdio' : 'unknown');
      const enabled = entry.enabled !== false;

      const dependencies: McpDependency[] = [];

      if (entry.command) {
        dependencies.push({ name: entry.command, met: checkCommand(entry.command) });
      } else if (entry.url) {
        const host = (() => { try { return new URL(entry.url).host; } catch { return entry.url; } })();
        dependencies.push({ name: host, met: checkHttpUrl(entry.url) });
      }

      const available = enabled && dependencies.every(d => d.met);

      return {
        kind: 'mcp' as const,
        id: `mcp#${name}`,
        name,
        url: entry.url,
        command: entry.command,
        args: entry.args,
        transport,
        headers: entry.headers,
        enabled,
        available,
        dependencies,
      };
    });
  } catch {
    return [];
  }
}

File: internal/data/mcp.go
package data

import (
	"encoding/json"
	"net/http"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

func checkCommand(cmd string) bool {
	_, err := exec.LookPath(cmd)
	return err == nil
}

func checkHTTPURL(url string) bool {
	client := &http.Client{Timeout: 3 * time.Second}
	resp, err := client.Head(url)
	if err != nil {
		return false
	}
	resp.Body.Close()
	return resp.StatusCode > 0
}

func LoadMCP() []OcMcpServer {
	home, _ := os.UserHomeDir()
	path := filepath.Join(home, ".openclaw", "openclaw.json")

	b, err := os.ReadFile(path)
	if err != nil {
		return nil
	}
	var cfg openclawJSON
	if err := json.Unmarshal(b, &cfg); err != nil {
		return nil
	}
	if cfg.MCP == nil {
		return nil
	}

	var items []OcMcpServer
	for name, entry := range cfg.MCP.Servers {
		transport := entry.Transport
		if transport == "" {
			if entry.Command != "" {
				transport = "stdio"
			} else {
				transport = "unknown"
			}
		}
		enabled := true
		if entry.Enabled != nil {
			enabled = *entry.Enabled
		}

		var deps []McpDependency
		if entry.Command != "" {
			deps = append(deps, McpDependency{Name: entry.Command, Met: checkCommand(entry.Command)})
		} else if entry.URL != "" {
			host := entry.URL
			// Extract just the host for display
			if idx := strings.Index(host, "://"); idx >= 0 {
				host = host[idx+3:]
				if idx2 := strings.IndexByte(host, '/'); idx2 >= 0 {
					host = host[:idx2]
				}
			}
			deps = append(deps, McpDependency{Name: host, Met: checkHTTPURL(entry.URL)})
		}

		available := enabled
		for _, d := range deps {
			if !d.Met {
				available = false
				break
			}
		}

		items = append(items, OcMcpServer{
			ID:           "mcp#" + name,
			Name_:        name,
			URL:          entry.URL,
			Command:      entry.Command,
			Args:         entry.Args,
			Transport:    transport,
			Headers:      entry.Headers,
			Enabled:      enabled,
			Available:    available,
			Dependencies: deps,
		})
	}
	return items
}
